New contract state

limemloh opened the pull request:
## Purpose

New state implementation for V1 contracts. This builds upon Concordium/concordium-wasm-smart-contracts#24 and
- Resolves Concordium/concordium-wasm-smart-contracts#29
- Resolves Concordium/concordium-wasm-smart-contracts#30
- Resolves Concordium/concordium-wasm-smart-contracts#39 

## Changes in wasm-chain-integration

- Provide a new implementation of contract state based on a radix tree with branching factor 16 (see Concordium/concordium-wasm-smart-contracts#38 for more details and rationale)
- Replace the state in V1 contracts with the new state implementation. This means changes in the validation of contracts to support new host functions, and implementation of handlers of host functions.
- For V1 contracts, keep track of whether the state has changed during execution. The definition of "state has changed" is whether any state modifying functions were called.
- Revised FFI for V1 contracts due to new handling of the state.

## Changes in cargo-concordium

- support for loading and storing the new state
- support for printing the new state, either at the end of execution, or as a standalone command

## Changes

Depends on:
- [x] https://github.com/Concordium/concordium-wasm-smart-contracts/pull/24
- [x] https://github.com/Concordium/concordium-contracts-common/pull/36

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

limemloh committed:
Implement V1 host instance state managed by FFI.

limemloh committed:
Include energy cost for host functions related to state.

limemloh committed:
Merge remote-tracking branch 'origin/sync-calls' into contract-state-trie-with-sync-calls

limemloh committed:
Fix iterator next FFI type

abizjak committed:
Merge branch 'sync-calls' into contract-state-trie-with-sync-calls

limemloh committed:
Remove tracking of entries and iterators

limemloh committed:
Include error handling in contract state FFI callbacks

limemloh committed:
Resume receive function always takes a new instance state

abizjak committed:
Merge branch 'sync-calls' into contract-state-trie-with-sync-calls

abizjak committed:
Merge branch 'sync-calls' into contract-state-trie-with-sync-calls

abizjak committed:
Use the new state implementation.

abizjak committed:
Introduce rust trie implementation.

abizjak committed:
Fix type of create entry.

abizjak committed:
Fix bug in freeze.

abizjak committed:
Collect the size of the trie while freezing it.

abizjak committed:
A basic test for size computation.

abizjak committed:
Better naming.

abizjak committed:
Tree serialization and accompanying tests.

MilkywayPirate committed:
.

MilkywayPirate committed:
..

MilkywayPirate committed:
.

MilkywayPirate committed:
comment

abizjak committed:
Merge branch 'main' into contract-state-trie-with-sync-calls

MilkywayPirate committed:
.

MilkywayPirate committed:
..

MilkywayPirate committed:
..

abizjak committed:
Fix invalidation in delete_subtree.

abizjak committed:
Some documentation.

MilkywayPirate committed:
.

MilkywayPirate committed:
Merge branch 'contract-state-iterators-and-invalidating-entries' of github.com:Concordium/concordium-wasm-smart-contracts into contract-state-iterators-and-invalidating-entries

MilkywayPirate committed:
.

MilkywayPirate committed:
refcount of locked nodes

MilkywayPirate committed:
locking of nodes

MilkywayPirate committed:
doc

MilkywayPirate committed:
doc

abizjak committed:
A bit of cleanup.

abizjak committed:
Merge branch 'contract-state-trie-with-sync-calls' into contract-state-iterators-and-invalidating-entries

MilkywayPirate committed:
alternative_locking wip

MilkywayPirate committed:
fix

MilkywayPirate committed:
a todo

MilkywayPirate committed:
fix

MilkywayPirate committed:
Prefix removal locking

abizjak committed:
Charge for children traversal when invalidating entries in the subtree.

MilkywayPirate committed:
Typos

Commit:
optimized delete_iter

Commit:
cleanup

abizjak committed:
Ensure there is no overflow when handing out iterators.

Commit:
.

Commit:
Merge branch 'alternative_locking' of github.com:Concordium/concordium-wasm-smart-contracts into alternative_locking

abizjak committed:
Make sure to not change node pointers.

The iterator has fixed node pointers. This commit makes it so that a node is
never updated so that pointers in the iterator remain valid upon insertion and
deletion in unlocked part of the tree.

abizjak committed:
Merge branch 'alternative_locking' into contract-state-iterators-and-invalidating-entries

abizjak committed:
Add tests with multiple iterators and prefixes.

abizjak committed:
Fix bug in generation management, add tests for generation locking.

abizjak committed:
Add energy accounting for most new operations.

abizjak committed:
Add a types module and split some data from the low_level into it.

Add documentation to some of the types that have been moved over.

abizjak committed:
Better naming for FlatLoadable/FlatStorable.

abizjak committed:
A bit more type safety around entry ids.

abizjak committed:
Revision and documentation of instance state implementation.

Some low_level methods needed to be revised to return informative enough messages.

abizjak committed:
Charge energy when advancing the iterator.

This cost is not constant and thus needs to be accounted for dynamically.

With more time we might be able to analyze and bound the cost, so that we could
be constant time per `next` call, but I do not have that insight now. However I
believe that should be possible.

abizjak committed:
Revise the host API, fix types in validation.

abizjak committed:
Refine return values, moving runtime errors into "error" responses for

- using entries that have been deleted, or that are invalid (e.g., wrong generation)
- accessing iterators that have been deleted.

abizjak committed:
Introduce an auxiliary trie to check for prefixes efficiently.

abizjak committed:
Iterator locking with a separate trie.

This merges
https://github.com/Concordium/concordium-wasm-smart-contracts/pull/33 but
significantly revises it - remove the generation from the iterator. It is not
needed, iterators and entries are always implicitly for the latest generation. -
use a trie to keep track of locked prefixes. This makes it siginificantly more
efficient to check whether a part of a tree is locked. It has O(key length)
complexity, regardless of how many iterators or locks there are. - add
documentation - fix tests

Co-authored-by: Emil Lai <el@concordium.com>"

abizjak committed:
Be more defensive with resizing entries.

abizjak committed:
Documentation, additional tests, and minor tweaks.

abizjak committed:
Revise state API

- Lock parts of the tree that has iterators
- Revise the exact return values of different methods to make them more uniform.

abizjak committed:
Enable cargo-concordium to use the new contract state.

abizjak committed:
Remove values from the values list when invalidating entries.

abizjak committed:
Add a lookup method via FFI. Fix serialization of host functions.

abizjak committed:
Expose via FFI a method for generating test states.

abizjak committed:
Keep track of state modifications and report this back through the FFI.

abizjak committed:
Revise semantics of the iterator get_key method.

abizjak committed:
Introduce fallback entrypoints.

abizjak committed:
Update dependencies to more recent versions.

abizjak committed:
Revise node branching to only use 4 bits for branching.

The reason for the change is that the worst case space usage for updates is
lessened.

With branching on all 8 bits the amount of new data that needs to be stored when
updating a value with key length `n` is n * (2338 + C) where C is a small
constant that determines how much data is needed to store a pointer to a
value (so C is around 9, depending on whether the value is present or not).
The 2338 comes from the following calculation. For each node on the way to the
root of the tree we need to store
2 bytes for the length of the children list
256 * (1 + 9) (1 for the key, 9 for the pointer to the child)
32 bytes for the hash of the node

With branching on 4 bits the computation instead yields n * (354 + 2C).

With branching on 2 bits we have even less at n * (276 + 4C) however the
difference is much less, and using fewer bits increases both the amount of
indirection, amount of wasted space, and amount of computation (more hashes to
compute). Thus 4 bits seems like the most reasonable value.

Commit:
Adds display-state sub-command to cargo concordium.

This command displays the contract state as a tree.

Also, adds a `--display-state` option to `run init` and `run update` subcommands.

abizjak committed:
Remove some obsolete todos.

abizjak committed:
Reduce visual noise with a helper function.

abizjak committed:
A bit of optimization.

abizjak committed:
Documentation.

MilkywayPirate committed:
Introduced tests for InstanceState.

MilkywayPirate committed:
Merge pull request #35 from Concordium/instance-state-tests

Tests of API

Commit:
Fixes clippy warnings

Commit:
Fixes the formatting problems found by `cargo fmt`

abizjak committed:
Fix bug in iterator's get_key method, and more extensive test for it.

Reported-by: Emil Lai <el@concordium.com>

abizjak committed:
Add benchmarks for V1 host functions, and revise costs accordingly.

Additionally some changes in the API.
- there are now 2^32 iterators allowed at any given prefix
- delete_entry API has changed. It now takes a key to delete, which alleviates
the need to keep the key in the host implementation, making cost assignment easier.

abizjak committed:
Merge branch 'contract-state-trie-with-sync-calls' of github.com:Concordium/concordium-wasm-smart-contracts into contract-state-trie-with-sync-calls

abizjak committed:
Reduce visual noise.

Commit:
Adds calling fallback entry point feature to `cargo-concordium`

When calling `cargo concordium run update` with an invalid entrypoint
the fallback entrypoint is called if one specified in the contract.

The name of the fallback entrypoint is assumed to be an empty string.

abizjak committed:
Simplify handling of fallback functions in cargo-concordium.

abizjak committed:
Fix bug in thaw.

abizjak committed:
Revise how "new state size" is computed.

This makes it so that when freezing we do nothing if the tree is not modified.
Additionally this fixes a bug in store_update for the node, where previously the
tree was not stored correctly.

abizjak committed:
Fix benchmarks after changes to freeze/thaw.

abizjak committed:
Documentation and minor refinements.

abizjak committed:
Bump submodule.

EmilLa1 commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3053 @@
+            CachedRef::Disk {
+                key,
+            } => {
+                let value = V::load_from_location(loader, *key).unwrap(); // TODO: Error handling.
```
Resolve now or later?

abizjak replied:
Added precondition.

MilkywayPirate reviewed (commented):
Looks good. 
A few questions and remarks of things we need to handle now or postpone.

abizjak committed:
Allow omission of policies in test receive contexts.

abizjak committed:
Rename version -> contract version, func -> entrypoint.

As agreed at a meeting yesterday.

abizjak commented on cargo-concordium/src/main.rs:
```diff
@@ -752,10 +803,23 @@ fn handle_run_v1(run_cmd: RunCommand, module: &[u8]) -> anyhow::Result<()> {
+     -> anyhow::Result<()> {
+        let mut collector = v1::trie::SizeCollector::default();
+        let frozen = state.freeze(loader, &mut collector);
+        // TODO: Decide on terminology to use for this so that it is clear what it
```
I revised the phrasing of the message.

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,304 @@
+    /// contract and released at the end.
+    /// The reason for the mutex is that we need to be able to clone this so
+    /// that we can share it inside the single transaction.
+    pub state: Arc<Mutex<MutableTrie<Value>>>,
```
Ok.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3053 @@
+/// representation where common parts of the key are stored inline in a single
+/// node instead of having many nodes with a single child.
+struct Stem {
+    // TODO: Split into short + long, or use smallvec/tinyvec, but make sure to benchmark any
```
Will make issue.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3053 @@
+/// A mutable version of the stem.
+/// This is an implementation detail of the iterator.
+struct MutStem {
+    // TODO: Split into short + long, or use smallvec/tinyvec, but make sure to benchmark any
```
Will make issue.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3053 @@
+}
+
+impl<V: Clone> Node<V> {
+    /// TODO: This is not very efficient. It involves cloning nodes, which is
```
Removed todo.

abizjak committed:
Add documentation, and revise some costs.

abizjak committed:
Merge branch 'contract-state-trie-with-sync-calls' of github.com:Concordium/concordium-wasm-smart-contracts into contract-state-trie-with-sync-calls

abizjak committed:
Add documentation, remove needless lifetimes.

EmilLa1 commented on wasm-chain-integration/src/v1/ffi.rs:
```diff
@@ -194,34 +218,61 @@ unsafe extern "C" fn call_receive_v1(
-        let receive_ctx = v0::deserial_receive_context(slice_from_c_bytes!(
+    let res = std::panic::catch_unwind(|| -> *mut u8 {
+        // For FFI we only pass v0 contexts to keep the other end simpler.
+        let receive_ctx_common = v0::deserial_receive_context(slice_from_c_bytes!(
```
Nevermind. I was just confused that v1 used the v0 namespace for deserializing. I do see now that the v0:ReceiveContext is common for both v0 and v1.

abizjak committed:
Ensure that parameters are bounded for V1 contracts as well.

abizjak committed:
Store node values inline.

abizjak committed:
Optimize storage a bit by using Box<[u8]> in place of Vec<u8> for immutable values.

abizjak committed:
Print the correct parameter for the send action.

abizjak committed:
Reduce max entry and key size to 1GB to allow use of 32-bit signed integers for seek.

abizjak committed:
Correctly mark nodes as modified in all cases of deletion.

Add regression test.

abizjak committed:
Check exports when building with cargo concordium.

abizjak committed:
Add test to make sure insertion and freezing behaves correctly.

abizjak committed:
Make the semantics of generations a tiny bit more strict.

Do not clear the top bit. This was not wrong per se, but was just a leftover
from the old encoding of options and no longer useful.

Bargsteen commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+/// This is used in this module to construct trees, allowing for sharing of
+/// values in trees and subtrees in case of the persistent tree.
+pub struct Link<V> {
+    link: Arc<RwLock<V>>,
```
Not sure I understand why the atomic part is needed and also the rwlock. Will revisit later, once I'm all the way through.

abizjak replied:
I have added this comment to attempt to explain why Arc and RwLock are needed.
```
/// This [Link] achieves the following properties
/// - it is cheap to clone
/// - it allows for inner mutability
/// - it is safe to use in a concurrent context.
///
/// The cheap cloning is necessary to have efficient persistent trees. The
/// [Link] is used to point to children, as well as values and in the persistent
/// tree, modifying the value at a given point means we need to create copies of
/// the spine of the tree up to the root. This involves cloning pointers to any
/// parts that have not been modified. Thus these have to be cheap, so we use an
/// [Arc].
///
/// Inner mutability is needed so that the persistent tree may be loaded from
/// disk, as well as written to disk. This is achieved in combination with
/// [CachedRef]. Instead of using an [RwLock] we could instead use a [Mutex]
/// which would achieve the same API. However based on benchmarks the RwLock has
/// negligible overhead in the case of a single reader, and thus the [RwLock]
/// seems the more natural choice since it allows concurrent reads of the tree.
///
/// Finally, the reference counting must be atomic since parts of the tree are
/// dropped concurrently. While all the operations of the smart contract
/// execution engine are sequential, multiple states may be derived from the
/// same state in different threads. Currently this happens when using
/// invokeContract, but in the future with parallel block execution it might
/// also happen during normal block processing.
/// Additionally, if the Haskell runtime is configured with the parallel garbage
/// collector then parts of the tree might be dropped concurrently. This also
/// requires atomic reference counting.
```

Bargsteen replied:
👌

Bargsteen reviewed (changes requested):
Still missing a deeper review of certain parts, but I'll post the comments now.

In general, I think the code looks good, but the big picture is hard to grasp.
I've been involved in some of the design decisions and overheard other comments about it, so I had some idea about how things were meant to work (fx the ideas behind freeze/thaw and generations), but I still found it hard to follow.

Some general documentation about the flow and design decisions would be greatly appreciated 😊 🙏  
- What happens when a contract entrypoint is invoked? What are the steps?
- What does a rollback look like and when does it happen?
- What is the idea behind the big concepts such as persistent/mutable state and generations.

Perhaps it could be explained from the standpoint of a typical flow with no optimizations and moving towards more optimizations, so you can see why the optimizations are there.

abizjak committed:
More consistent naming and some documentation.

abizjak commented on wasm-chain-integration/src/constants.rs:
```diff
@@ -40,15 +40,15 @@ pub const BASE_SIMPLE_TRANSFER_ACTION_COST: u64 = BASE_ACTION_COST + 40000;
 #[inline(always)]
 pub fn copy_from_host_cost(x: u32) -> u64 { 10 + u64::from(x) }
-/// Cost of copying the given amount of bytes from the host (e.g., parameter or
+/// Cost of copying the given amount of bytes to the host (e.g., parameter or
```
I have added some more documentation. I hope that reading this additional documentation gives enough insight into how the costs are determined, even if this is not explicit for each and every cost.

abizjak committed:
Rename callback types for consistency with Haskell code.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+    let value_mask: u8 = if no_value {
+        0
+    } else {
+        0b0100_0000
```
It is not so much that they are fixed relative to that, it is that "INLINE_STEM_LENGTH" has to fit into 6 bits.

These masks are just test the first two bits. I feel that introducing constants would only make it harder to read.

I will attempt to provide a comment where the tags are used.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+            }
+            Ok(())
+        };
+        let mut ref_stack = Vec::<Reference>::new();
```
I have attempted to explain it.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+                        buf.write_all(&v[1..][0..usize::from(v[0])])?;
+                    }
+                    InlineOrHashed::Indirect(indirect) => {
+                        buf.write_u8(0b1111_1111u8)?;
```
It is now.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+                let mut borrowed = v.borrow_mut();
+                match &mut *borrowed {
+                    InlineOrHashed::Inline(v) => {
+                        buf.write_u8(v[0])?;
```
It is documented now. I do not want to add an explicit length field since it would increase the variant by 8 bytes which is not insignificant.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+    /// if it is not yet cached. Note that this will **not** cache the
+    /// value, the loaded value will be dropped.
+    #[inline(always)]
+    pub(crate) fn use_value<X>(
```
Done.

abizjak committed:
Documentation of some constants and the store_update algorithm.

abizjak committed:
Replace the use of callbacks in use_value with get_ref for a more idiomatic Rust API.

abizjak committed:
More consistent naming of lookup methods.

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,308 @@
+            PersistentState::Empty => out.write_u8(0)?,
+            PersistentState::Root(ht) => {
+                out.write_u8(1)?;
+                ht.use_value_(loader, |loader, node| node.serialize(loader, out))?
```
I agree, I just did not know what the better name is. This pattern that I'm using here is not very rust like. It is very javascripty or Haskelly with supplying a callback to do stuff.

Instead of renaming, I have changed these `use_value` methods to `.get` methods that return something like `Cow` (but not quite that due to needless constraints of that type).

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+/// A persistent node. Cloning this is cheap, it only copies pointers and
+/// increments reference counts.
+pub struct Node {
+    /// Since a single node owns each value using Hashed<Cached<V>>
+    /// here makes sense, it makes it so that the hash is stored inline.
```
Indeed it is. Obsolete comments mixing various previous versions of the structure :)

abizjak commented on wasm-chain-integration/src/v1/trie/tests.rs:
```diff
@@ -0,0 +1,1027 @@
+        };
+        let mut ser = Vec::new();
+        let _ = frozen.store_update(&mut ser);
+        ensure!(frozen.get(&mut loader).data.is_stored(), "Not all data is stored.");
```
Indeed.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+}
+
+/// The default hash implementation is not a valid value.
+impl<V> Default for CachedRef<V> {
```
It was used in the `deserialize` method via the `std::mem::take` method. I have changed that to use `mem::replace` instead with a more explicit dummy value, and removed the default.

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,308 @@
+    /// contract and released at the end.
+    /// The reason for the mutex is that we need to be able to clone this so
+    /// that we can share it inside the single transaction.
+    state: Arc<Mutex<MutableTrie>>,
```
We need to be able to mutate the state, so we need the mutex. If we just had `Arc<MutableTrie>` then it is never possible to get a mutable reference to `MutableTrie`, and we need that.

`Arc<Mutex<T>>` is the equivalent to `Rc<RefCell<T>>`.

While it is true that execution is single-threaded the general execution context is multithreaded since (in general at least) we are using a parallel garbage collector. 

In this specific case I think it might be possible to use `Rc<RefCell<T>>` instead, but we'd need to carefully analyze that we don't have issues with cleanup (which is triggered via the Haskell GC in some cases). Since this does not impact performance I don't think the risks and time spent analyzing to make sure it would work are worth it.

We need multiple copies of this due to generations and rollback.

abizjak committed:
Improve documentation and naming.

abizjak committed:
Document reasons for data structure choice.

Rename key -> reference to avoid confusion.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+                value,
+            } => {
+                let key = backing_store.store_raw(value.as_ref())?;
+                let value = std::mem::replace(self, CachedRef::Disk {
```
Ok.

This is not the most ergonomic part of rust.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+/// and a pointer to the actual value. The intention is that small values will
+/// be stored inline and larger values behind and indirection.
+pub enum InlineOrHashed {
+    Inline([u8; INLINE_VALUE_LEN + 1]),
```
Ok.

The reason inlining into the byte array is to make sure no extra space is used. Due to alignment and cleverness of the compiler there seems to be no difference in the size of the type when factoring the value out. But it is more implicit.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+}
+
+impl Node {
+    /// The entire tree in memory.
```
I chose to elaborate a bit more.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+        Ok(buf)
+    }
+
+    pub fn store_update_buf<S: BackingStoreStore, W: std::io::Write>(
```
The problem with stack is that it is very implicit, and limited (by default I believe 8MB per thread on my machines at least, not sure about other platforms). This means it is very hard to argue for correctness in cases where the inputs (the tree) can be constructed maliciously. Another issue is that it matters what the call stack is before this function is called, so the available stack space for recursive calls might actually be much less.

Using an explicit stack (with vectors) is much more predicable, explicit, and safe, i.e., we can guarantee that we don't have to worry about that limit.

Regarding performance, I would not expect that using an explicit stack (i.e., without recursion in the language) would be faster.

I agree that it is harder to follow, but it needs to work. I also think that once you get a hang of rewriting recursive functions into functions without recursion and explicit control stack this is a fairly mechanical transformation.

abizjak replied:
In any case, there is more documentation of the approach, as well as details in this function.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+        };
+    }
+
+    pub fn next<L: BackingStoreLoad, C: TraversalCounter>(
```
Done.

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,308 @@
+    pub fn make_fresh_generation(&mut self, loader: &mut impl BackingStoreLoad) -> Self {
+        if let Some(inner) = self.inner.as_mut() {
+            let mut trie = inner.lock();
+            trie.normalize(inner.root);
```
I am not sure I understand this question. Since the MutableTrie is shared, during rollback we do not actually clean it up.

abizjak committed:
Record length separately for clarity.

abizjak commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -0,0 +1,445 @@
+impl<'a, A: AsRef<[u8]>> BackingStoreLoad for Loader<A> {
+    type R = Vec<u8>;
+
+    // FIXME: This is inefficient. We allocate too many vectors.
```
Resolved by using tinyvec

abizjak commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -0,0 +1,445 @@
+    pub fn collect(self) -> u64 { self.num_bytes }
+}
+
+// TODO: Make sure this is adequate. There is a bit of overhead with size length
```
Now.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3053 @@
+        }
+    }
+
+    /// TODO: It might be useful to return a list of new nodes so that they
```
I elaborated a bit for future maintainers.

abizjak commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -0,0 +1,445 @@
+/// A type that can be used to collect auxiliary information while a mutable
+/// trie is being frozen. Particular use-cases of this are collecting the size
+/// of new data, as well as new persistent nodes.
+/// TODO: Methods should probably return Result/Option so that we can terminate
```
I explained what could be done, and that what we do now is not too much of an issue.

If we have time in the future we can do it.

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,308 @@
+//! The high-level API that is meant to be used by the rest of the project.
+//! It exposes two main constructs, the [PersistentState] and [MutableState]
```
I have added a few paragraphs.
```rust

//! The [`PersistentState`] is immutable. It is "persistent" in the sense of
//! persistent data structures, meaning that we never modify it in-place, but
//! always create new versions. This state is used for recording the state of a
//! smart contract at the end of a transaction execution.
//!
//! State modifications are done via the [`MutableState`] which exists during
//! transaction execution. This type is modified in-place and supports rollbacks
//! which are sometimes necessary during the transaction due to contract calls.
//!
//! The overall flow is that a [`PersistentState`] is
//! [thawed](PersistentState::thaw) into a [`MutableState`] when the contract
//! starts execution. Then a contract is executed, and if it terminates with
//! success the resulting state is [frozen](MutableState::freeze) into a
//! [`PersistentState`]. During this process only the part of the original state
//! that has been modified is copied. The new [`PersistentState`] in general
//! retains pointers to parts of the original state. This allows for efficient
//! state updates where we only have to store the parts of the state that are
//! new.
```

abizjak commented on wasm-chain-integration/src/v1/trie/api.rs:
```diff
@@ -0,0 +1,308 @@
+/// copy**. Modifications of the inner MutableTrie on either the clone or the
+/// original propagate to the other.
+pub struct MutableStateInner {
+    /// Current root of the tree. The current generation.
```
I have tried to explain it now.

abizjak commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -0,0 +1,471 @@
+/// TODO: Methods should probably return Result/Option so that we can terminate
+/// early in case of out of energy.
+pub trait Collector<V> {
+    fn add_value(&mut self, data: &V);
```
Done.

abizjak commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -0,0 +1,471 @@
+/// make sure that resource bounds are not exceeded.
+pub trait TraversalCounter {
+    type Err: std::fmt::Debug;
+    fn tick(&mut self, num: u64) -> Result<(), Self::Err>;
```
Added (and also renamed not to confused with the energy `.tick`).

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+    }
+
+    /// Extend the current stem by the content of the given stem.
+    fn extend(&mut self, second: &Stem) {
```
Added.

abizjak commented on wasm-chain-integration/src/v1/mod.rs:
```diff
@@ -315,12 +353,280 @@ mod host {
+        energy: &mut InterpreterEnergy,
+        state: &mut InstanceState<BackingStore>,
+    ) -> machine::RunResult<()> {
+        let key_len = unsafe { stack.pop_u32() };
+        let key_start = unsafe { stack.pop_u32() } as usize;
+        let key_end = key_start + key_len as usize;
+        // this cannot overflow on 64-bit platforms, so it is safe to just add
+        ensure!(key_end <= memory.len(), "Illegal memory access.");
+        let key = &memory[key_start..key_end];
```
I prefer not to do that. All of these methods are structured in a similar way with arguments being parsed inline.

I do not feel that factoring out this part would bring much value.

abizjak commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3164 @@
+}
+
+impl MutableTrie {
+    pub fn new_generation(&mut self) {
```
> Also on how it relates to `normalize`. Should `normalize` always be called prior to `new_generation`?

No. Normalize just forgets 0 or more of the most recent generations. I have added some documentation that hopefully explains this.

abizjak replied:
I have added more documentation both to the api module and the low_level module outlining the overall approach.

abizjak committed:
Add more documentation describing the overall approach.

abizjak commented:
@Bargsteen I have added a bunch more documentation now. I hope some of the high-level concepts are clearer now.

There could be more detailed documentation, but I need to focus on merging this and the node changes in. So let me know if there is anything that is critically missing.

abizjak committed:
Merge remote-tracking branch 'origin/main' into contract-state-trie-with-sync-calls

abizjak committed:
Mark state as changed.

Bargsteen commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3391 @@
+/// - it is safe to use in a concurrent context.
+///
+/// The cheap cloning is necessary to have efficient persistent trees. The
+/// [Link] is used to point to children, as well as values and in the persistent
```
```suggestion
/// [Link] is used to point to children, as well as values, in the persistent
```

Bargsteen commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3391 @@
+///
+/// The cheap cloning is necessary to have efficient persistent trees. The
+/// [Link] is used to point to children, as well as values and in the persistent
+/// tree, modifying the value at a given point means we need to create copies of
```
```suggestion
/// tree. Modifying the value at a given point means we need to create copies of
```

Bargsteen commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -0,0 +1,3391 @@
+/// reader, and thus the [RwLock] seems the more natural choice since it allows
+/// concurrent reads of the tree.
+///
+/// Finally, the reference counting must be atomic since parts of the tree are
+/// dropped concurrently. While all the operations of the smart contract
+/// execution engine are sequential, multiple states may be derived from the
+/// same state in different threads. Currently this happens when using
+/// invokeContract, but in the future with parallel block execution it might
+/// also happen during normal block processing.
+/// Additionally, if the Haskell runtime is configured with the parallel garbage
+/// collector then parts of the tree might be dropped concurrently. This also
+/// requires atomic reference counting.
```
👌

Bargsteen commented on wasm-chain-integration/src/v1/trie/low_level.rs:
```diff
@@ -1382,15 +1515,34 @@ impl Node {
+        // - recursively write all the children that are only in memory
+        // - this would return Reference's to where these children are written in the
+        //   backing store
+        // - use these to write the node into the provided buffer (`buf`).
```
:ok_hand:

Bargsteen commented on wasm-chain-integration/src/v1/trie/types.rs:
```diff
@@ -53,14 +53,17 @@ pub type LoadResult<A> = Result<A, LoadError>;
 /// Length of the stem that will be stored inline, i.e.,
 /// the stem length will be encoded in the tag bit, as opposed to separate
-/// 4 bytes.
+/// 4 bytes. This must always fit 6 bits, i.e., can be no more than 127.
```
63?
```suggestion
/// 4 bytes. This must always fit 6 bits, i.e., can be no more than 63.
```

Bargsteen reviewed (approved):
The added documentation was much appreciated 🎉 
LGTM 👍

abizjak committed:
Revise cargo-concordium changelog.

abizjak committed:
Fix typos.
